v2/sourcedb-to-spanner/terraform/samples/sharded-bulk-migration/README.md

```diff
@@ -0,0 +1,418 @@
+
+**Note 2:** Terraform, by default, creates at most 10 resources in parallel. If you want to run more than 10 dataflow
+jobs in parallel, run terraform using the `-parallelism=n` flag. However, more parallel dataflow jobs linearly increases
+load on Spanner.
+If Spanner is not sized sufficiently, all the jobs will slow down.
+
+```shell
+terraform apply -parallelism=30
+```
+
+### Cleanup
+
+Once the jobs have finished running, you can cleanup by running -
+
+```shell
+terraform destroy
+```
+
+## FAQ
+
+### Dataflow job is failing with "Timeout in polling result file"
+
+Dataflow has a 10min timeout within which the launcher VM logic should complete. There could be multiple reasons for it
+to take over 10 mins:
+
+- **Job logs not present after the log "launcher VM started":** A sign would be there are only 3-4 log statements in the
+  job logs. This likely due to private google access is not enabled for the
+  subnetwork. Please enable private google access in your network.
+- **Job logs not present after "Discovering tables for DataSource":** This means dataflow is unable to access the mysql
+  shard. Please check your network configuration and credentials.
+- If you are still getting a Timeout after this, this means the job graph construction is taking too long. You could
+  potentially have too many tables allocated in a single Dataflow job. The recommendation is to set the `batch_size`
+  such that the total number of tables in a single job does not exceed `150`. This means if a logical shard has 15
+  tables, and a physical shard has 2 logical shards, do not set the batch_size more than 5.
+
+### Job graph is not loading/Custom counters not visible on Dataflow panel
+
+For very large graphs, this can happen. The graph section would be empty and the counters won't load. But worry not, the
+migration should progress nonetheless.
+
+### Data being written to Spanner is too slow
+
+THere can be multiple reasons for this.
+
+- **Check source shard metrics:** Are memory/CPU limits being hit? Consider increasing the shard resources.
+- **Check Spanner metrics:** Are memory/CPU limits being hit? Consider increasing the number of nodes.
+- **Check Dataflow metrics:** Are memory/CPU limits being hit? Dataflow should autoscale to required number of
+  nodes.
+    - CPU/memory limits being hit means the `max-workers` parameter might be too low. It is recommended to use smaller
+      machines (`Ex: n1-highmem-4`) for most workloads. However, if you have tall tables (singe table larger than 1tb),
+      we recommend using larger machine sizes (`n1-highmem-32`).
+    - For tables with `string` primary keys, the throughput is expected to be lower than tables with `int` primary keys.
+    - Consider bumping up `maxConnection` and `numPartition` parameters when launching the jobs. We recommend `320` for
+      maxConnections and `10000` for numPartitions.
+
+### Configuring to run using a VPC
```

Comment:
Are there any firewall rules that need to be configured in a VPC to get Dataflow VMs talking to each and the source database VMs?

Reply:
Updated docs